File: Leetcode 3701-3800/3714/3714.py
# Leetcode 3714: Longest Balanced Substring II
# https://leetcode.com/problems/longest-balanced-substring-ii/
# Solved on 4th of November, 2025
class Solution:
    def longestBalanced(self, s: str) -> int:
        """
        Finds the length of the longest balanced substring in a given string.

        Args:
            s (str): The input string consisting of characters 'a', 'b', and 'c'.
        Returns:
            int: The length of the longest balanced substring.
        """
        maxLength = 0

        def solveOne(c1, c2, c3):
            length = 0
            currentLength = 0
            for char in s:
                if char == c1:
                    currentLength += 1
                elif char == c2 or char == c3:
                    length = max(length, currentLength)
                    currentLength = 0
            length = max(length, currentLength)
            return length

        def solveTwo(c1, c2, c3):
            counts = {c1: 0, c2: 0}
            seen = {0: -1}
            length = 0
            for i, char in enumerate(s):
                if char == c1:
                    counts[c1] += 1
                elif char == c2:
                    counts[c2] += 1
                elif char == c3:
                    counts = {c1: 0, c2: 0}
                    seen = {0: i}
                    continue
                else:
                    continue

                diff = counts[c1] - counts[c2]
                if diff in seen:
                    length = max(length, i - seen[diff])
                else:
                    seen[diff] = i
            return length

        def solveThree():
            seen = {(0, 0): -1}
            length = 0
            countA = 0
            countB = 0
            countC = 0
            for i, char in enumerate(s):
                if char == 'a':
                    countA += 1
                elif char == 'b':
                    countB += 1
                elif char == 'c':
                    countC += 1

                key = (countA - countB, countB - countC)

                if key in seen:
                    length = max(length, i - seen[key])
                else:
                    seen[key] = i
            return length

        maxLength = max(maxLength, solveOne('a', 'b', 'c'))
        maxLength = max(maxLength, solveOne('b', 'a', 'c'))
        maxLength = max(maxLength, solveOne('c', 'a', 'b'))
        maxLength = max(maxLength, solveTwo('a', 'b', 'c'))
        maxLength = max(maxLength, solveTwo('a', 'c', 'b'))
        maxLength = max(maxLength, solveTwo('b', 'c', 'a'))
        maxLength = max(maxLength, solveThree())

        return maxLength

File: Leetcode 3701-3800/3714/3714.java
// Leetcode 3714: Longest Balanced Substring II
// https://leetcode.com/problems/longest-balanced-substring-ii/
// Solved on 4th of November, 2025
import java.util.*;

class Solution {
    
    /**
     * Finds the length of the longest balanced substring in a given string.
     * @param s The input string consisting of characters 'a', 'b', and 'c'.
     * @return The length of the longest balanced substring.
     */
    public int longestBalanced(String s) {
        int n = s.length();
        if (n == 0) return 0;
        int maxLen = 1;
        int runLen = 1;
        for (int i = 1; i < n; i++) {
            if (s.charAt(i) == s.charAt(i - 1)) {
                runLen++;
            } else {
                if (runLen > maxLen) maxLen = runLen;
                runLen = 1;
            }
        }
        if (runLen > maxLen) maxLen = runLen;

        int size = 2 * n + 5;
        int offset = n + 2;
        char[] chars = {'a', 'b', 'c'};
        for (int p = 0; p < 3; p++) {
            char x = chars[p];
            char y = chars[(p + 1) % 3];
            char z = chars[(p + 2) % 3];
            int[] seenIndex = new int[size];
            int[] seenToken = new int[size];
            int token = 1;
            int base = -1;
            int diff = 0;
            seenToken[offset] = token;
            seenIndex[offset] = base;
            for (int i = 0; i < n; i++) {
                char c = s.charAt(i);
                if (c == z) {
                    token++;
                    base = i;
                    diff = 0;
                    seenToken[offset] = token;
                    seenIndex[offset] = base;
                } else {
                    if (c == x) diff++;
                    else if (c == y) diff--;
                    int idx = diff + offset;
                    if (idx < 0) {
                        continue;
                    }
                    if (idx >= size) {
                        continue;
                    }
                    if (seenToken[idx] != token) {
                        seenToken[idx] = token;
                        seenIndex[idx] = i;
                    } else {
                        int len = i - seenIndex[idx];
                        if (len > maxLen) maxLen = len;
                    }
                }
            }
        }

        HashMap<Long, Integer> tripleMap = new HashMap<>(Math.max(16, n * 2));
        int aCount = 0;
        int bCount = 0;
        int cCount = 0;
        tripleMap.put(0L, -1);
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            if (ch == 'a') aCount++;
            else if (ch == 'b') bCount++;
            else cCount++;
            int diff1 = aCount - bCount;
            int diff2 = aCount - cCount;
            long key = (((long)diff1) << 32) ^ (diff2 & 0xffffffffL);
            Integer first = tripleMap.get(key);
            if (first == null) tripleMap.put(key, i);
            else {
                int len = i - first;
                if (len > maxLen) maxLen = len;
            }
        }
        return maxLen;
    }
}